File: Backend/Ajial.Application/Service/AnalyticsService.cs
using Ajial.Application.DTOs.Analytics;
using Ajial.Application.DTOs.Common;
using Ajial.Application.Interfaces;
using Ajlal.Application.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Ajial.Application.Services;

public class AnalyticsService : IAnalyticsService
{
    private readonly IUnitOfWork _unitOfWork;

    public AnalyticsService(IUnitOfWork unitOfWork)
    {
        _unitOfWork = unitOfWork;
    }

    public async Task<ApiResponse<List<ChildAnalyticsDto>>> GetAllChildrenAnalyticsAsync()
    {
        try
        {
            // Get all children with related data
            var children = await _unitOfWork.Children.GetAllAsync(query => 
                query.Include(c => c.Parent)
                     .ThenInclude(p => p.User)
                     .Include(c => c.Parent)
                     .ThenInclude(p => p.City)
            );

            var result = children.Select(child => new ChildAnalyticsDto
            {
                // Child Information
                ChildId = child.Id,
                ChildFullName = child.FullName,
                ChildBirthDate = child.BirthDate,
                ChildAge = child.Age,
                ChildGender = child.Gender,
                ChildProfileImageUrl = child.ProfileImageUrl,
                ChildLoginId = child.ChildLoginId,
                ChildHasAccount = !string.IsNullOrEmpty(child.ChildLoginId),
                ChildIsActive = child.IsActive,
                ChildCreatedAt = child.CreatedAt,
                ChildUpdatedAt = child.UpdatedAt,
                
                // Parent Information
                ParentId = child.Parent.Id,
                ParentUserId = child.Parent.UserId,
                ParentFullName = child.Parent.User.FullName,
                ParentUsername = child.Parent.User.Username,
                ParentEmail = child.Parent.User.Email,
                ParentGender = child.Parent.Gender.ToString(),
                ParentDateOfBirth = child.Parent.DateOfBirth,
                ParentAge = CalculateAge(child.Parent.DateOfBirth),
                ParentIsActive = child.Parent.User.IsActive,
                ParentCreatedAt = child.Parent.CreatedAt,
                
                // City Information
                CityId = child.Parent.CityId,
                CityName = child.Parent.City.Name,
                CityNameAr = child.Parent.City.NameAr,
                
                // Calculated Fields
                AgeGroup = GetAgeGroup(child.Age),
                DaysSinceRegistration = (DateTime.UtcNow - child.CreatedAt).Days,
                RegistrationMonth = child.CreatedAt.ToString("yyyy-MM"),
                RegistrationYear = child.CreatedAt.Year.ToString()
            }).ToList();

            return ApiResponse<List<ChildAnalyticsDto>>.SuccessResponse(
                result,
                $"تم جلب بيانات {result.Count} طفل بنجاح"
            );
        }
        catch (Exception ex)
        {
            return ApiResponse<List<ChildAnalyticsDto>>.FailureResponse(
                "حدث خطأ أثناء جلب بيانات الأطفال",
                new List<string> { ex.Message }
            );
        }
    }

    public async Task<ApiResponse<List<ParentChildrenFlatDto>>> GetParentsWithChildrenFlatAsync()
    {
        try
        {
            // Get all parents with their children
            var parents = await _unitOfWork.Parents.GetAllAsync(query =>
                query.Include(p => p.User)
                     .Include(p => p.City)
                     .Include(p => p.Children)
            );

            var result = new List<ParentChildrenFlatDto>();

            foreach (var parent in parents)
            {
                // Calculate parent statistics
                var totalChildren = parent.Children.Count;
                var activeChildren = parent.Children.Count(c => c.IsActive);
                var childrenWithAccounts = parent.Children.Count(c => !string.IsNullOrEmpty(c.ChildLoginId));

                if (parent.Children.Any())
                {
                    // Create one row per child
                    foreach (var child in parent.Children)
                    {
                        result.Add(new ParentChildrenFlatDto
                        {
                            // Parent Information
                            ParentId = parent.Id,
                            ParentUserId = parent.UserId,
                            ParentFullName = parent.User.FullName,
                            ParentUsername = parent.User.Username,
                            ParentEmail = parent.User.Email,
                            ParentGender = parent.Gender.ToString(),
                            ParentDateOfBirth = parent.DateOfBirth,
                            ParentAge = CalculateAge(parent.DateOfBirth),
                            ParentIsActive = parent.User.IsActive,
                            ParentCreatedAt = parent.CreatedAt,
                            
                            // City Information
                            CityId = parent.CityId,
                            CityName = parent.City.Name,
                            CityNameAr = parent.City.NameAr,
                            
                            // Parent Statistics
                            TotalChildren = totalChildren,
                            ActiveChildren = activeChildren,
                            ChildrenWithAccounts = childrenWithAccounts,
                            
                            // Child Information
                            ChildId = child.Id,
                            ChildFullName = child.FullName,
                            ChildBirthDate = child.BirthDate,
                            ChildAge = child.Age,
                            ChildGender = child.Gender,
                            ChildProfileImageUrl = child.ProfileImageUrl,
                            ChildLoginId = child.ChildLoginId,
                            ChildHasAccount = !string.IsNullOrEmpty(child.ChildLoginId),
                            ChildIsActive = child.IsActive,
                            ChildCreatedAt = child.CreatedAt,
                            ChildUpdatedAt = child.UpdatedAt,
                            
                            // Child Calculated Fields
                            ChildAgeGroup = GetAgeGroup(child.Age),
                            ChildDaysSinceRegistration = (DateTime.UtcNow - child.CreatedAt).Days
                        });
                    }
                }
                else
                {
                    // Parent with no children - still include one row
                    result.Add(new ParentChildrenFlatDto
                    {
                        // Parent Information
                        ParentId = parent.Id,
                        ParentUserId = parent.UserId,
                        ParentFullName = parent.User.FullName,
                        ParentUsername = parent.User.Username,
                        ParentEmail = parent.User.Email,
                        ParentGender = parent.Gender.ToString(),
                        ParentDateOfBirth = parent.DateOfBirth,
                        ParentAge = CalculateAge(parent.DateOfBirth),
                        ParentIsActive = parent.User.IsActive,
                        ParentCreatedAt = parent.CreatedAt,
                        
                        // City Information
                        CityId = parent.CityId,
                        CityName = parent.City.Name,
                        CityNameAr = parent.City.NameAr,
                        
                        // Parent Statistics
                        TotalChildren = 0,
                        ActiveChildren = 0,
                        ChildrenWithAccounts = 0,
                        
                        // No child data
                        ChildId = null,
                        ChildFullName = null,
                        ChildBirthDate = null,
                        ChildAge = null,
                        ChildGender = null,
                        ChildProfileImageUrl = null,
                        ChildLoginId = null,
                        ChildHasAccount = null,
                        ChildIsActive = null,
                        ChildCreatedAt = null,
                        ChildUpdatedAt = null,
                        ChildAgeGroup = null,
                        ChildDaysSinceRegistration = null
                    });
                }
            }

            return ApiResponse<List<ParentChildrenFlatDto>>.SuccessResponse(
                result,
                $"تم جلب بيانات {parents.Count()} ولي أمر و {result.Count(r => r.ChildId != null)} طفل"
            );
        }
        catch (Exception ex)
        {
            return ApiResponse<List<ParentChildrenFlatDto>>.FailureResponse(
                "حدث خطأ أثناء جلب بيانات أولياء الأمور والأطفال",
                new List<string> { ex.Message }
            );
        }
    }

    private int CalculateAge(DateTime birthDate)
    {
        var today = DateTime.UtcNow;
        var age = today.Year - birthDate.Year;
        if (birthDate.Date > today.AddYears(-age)) age--;
        return age;
    }

    private string GetAgeGroup(int age)
    {
        return age switch
        {
            <= 3 => "0-3",
            <= 7 => "4-7",
            <= 13 => "8-13",
            _ => "14+"
        };
    }
}

File: Admin/ajial-admin/src/app/services/analytics.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Child {
  childId: string;
  childFullName: string;
  childBirthDate: string;
  childAge: number;
  childGender: string;
  childProfileImageUrl: string;
  childLoginId: string | null;
  childHasAccount: boolean;
  childIsActive: boolean;
  childCreatedAt: string;
  childUpdatedAt: string;
  parentId: string;
  parentUserId: string;
  parentFullName: string;
  parentUsername: string;
  parentEmail: string;
  parentGender: string;
  parentDateOfBirth: string;
  parentAge: number;
  parentIsActive: boolean;
  parentCreatedAt: string;
  cityId: number;
  cityName: string;
  cityNameAr: string;
  ageGroup: string;
  daysSinceRegistration: number;
  registrationMonth: string;
  registrationYear: string;
}

export interface Parent {
  userId: string;
  fullName: string;
  username: string;
  email: string;
  userCreatedAt: string;
  userUpdatedAt: string | null;
  parentId: string;
  dateOfBirth: string;
  age: number;
  gender: string;
  genderCode: number;
  cityId: number;
  cityName: string;
  cityNameAr: string;
  cityIsActive: boolean;
  accountAgeInDays: number;
  ageGroup: string;
  region: string;
}

export interface ParentSummary {
  totalParents: number;
  maleCount: number;
  femaleCount: number;
  averageAge: number;
  ageGroups: { ageGroup: string; count: number }[];
  citiesDistribution: { city: string; count: number }[];
  regionsDistribution: { region: string; count: number }[];
  recentRegistrations: number;
  lastUpdated: string;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  errors: any;
}

@Injectable({
  providedIn: 'root'
})
export class AnalyticsService {
  private http = inject(HttpClient);
  private baseUrl = 'https://ajial-api-dev-dvg9hfgtdgewekcv.westeurope-01.azurewebsites.net/api';

  getChildrenAnalytics(): Observable<ApiResponse<Child[]>> {
    return this.http.get<ApiResponse<Child[]>>(`${this.baseUrl}/Analytics/children`);
  }

  getParentsAnalytics(): Observable<ApiResponse<Parent[]>> {
    return this.http.get<ApiResponse<Parent[]>>(`${this.baseUrl}/Parents/analytics`);
  }

  getParentsSummary(): Observable<ApiResponse<ParentSummary>> {
    return this.http.get<ApiResponse<ParentSummary>>(`${this.baseUrl}/Parents/analytics/summary`);
  }

  getChildProfileSummary(childId: string): Observable<ApiResponse<any>> {
    return this.http.get<ApiResponse<any>>(`${this.baseUrl}/Parents/child/${childId}/profile-summary`);
  }
}
